File: src/app/models/races/DlsDeclaration.ts
/**
 * DLS Declaration model
 * Represents a user's declaration that they intend to DLS (Dead Last Start) a race
 */
export interface DlsDeclaration {
	/** Declaration ID */
	id: number;
	/** The DLS race this declaration belongs to */
	dlsRaceId: number;
	/** Display name of the DLS race */
	dlsRaceName: string;
	/** The runner's bib number (optional, may be added later) */
	bibNumber: number | null;
	/** The user ID who made this declaration */
	userId: number | null;
	/** Display name of the user */
	userDisplayName: string | null;
	/** Whether this is the user's first DLS */
	isFirstDls: boolean;
	/** Whether the user is going for kills */
	isGoingForKills: boolean;
	/** Optional comments from the user */
	comments: string | null;
	/** When this declaration was created */
	createdAt: string;
}

/**
 * Request to create a new DLS race (admin)
 */
export interface CreateDlsRaceRequest {
	/** Display name for the DLS race */
	name: string;
	/** The date of the race */
	raceDate: string;
}

/**
 * Request to update a DLS race (admin)
 */
export interface UpdateDlsRaceRequest {
	/** Updated name (optional) */
	name?: string;
	/** Updated date (optional) */
	raceDate?: string;
}

/**
 * Request to self-declare for a DLS race
 */
export interface CreateDlsDeclarationRequest {
	/** The DLS race to declare for */
	dlsRaceId: number;
	/** Optional bib number */
	bibNumber?: number;
	/** Whether this is the user's first DLS */
	isFirstDls?: boolean;
	/** Whether the user is going for kills */
	isGoingForKills?: boolean;
	/** Optional comments */
	comments?: string;
}

/**
 * Request to update a declaration
 */
export interface UpdateDlsDeclarationRequest {
	/** Updated bib number */
	bibNumber?: number | null;
	/** Whether this is the user's first DLS */
	isFirstDls?: boolean;
	/** Whether the user is going for kills */
	isGoingForKills?: boolean;
	/** Optional comments */
	comments?: string;
}

export interface ImportDlsDeclarationRequest {
	/** The user's display name (for imported declarations without a user account - attempt to match to an account, otherwise ignored) */
	name?: string;

	/** Bib number is a required field for imports to allow matching to existing declarations and users. Admins can provide this when importing, users can add later if self-declaring without a bib. */
	bibNumber: number;

	/** Whether this is the user's first DLS */
	isFirstDls?: boolean;

	/** Whether the user is going for kills */
	isGoingForKills?: boolean;

	/** Optional user comments */
	comments?: string;
}

File: src/api/Falchion.Villains.Vault.Api/Data/Entities/DlsDeclaration.cs
namespace Falchion.Villains.Vault.Api.Data.Entities;

/// <summary>
/// Represents a user's declaration that they intend to Dead Last Start (DLS) an upcoming race.
/// Declarations can be created by admins (with bib numbers for known runners) or by users
/// self-declaring (with optional bib number added later).
/// </summary>
public class DlsDeclaration
{
	/// <summary>
	/// Primary key - auto-incrementing integer ID.
	/// </summary>
	public int Id { get; set; }

	/// <summary>
	/// Foreign key to the DLS race entry.
	/// </summary>
	public int DlsRaceId { get; set; }

	/// <summary>
	/// The runner's bib number, if known.
	/// May be null if the user self-declared but doesn't have their bib yet.
	/// </summary>
	public int? BibNumber { get; set; }

	/// <summary>
	/// Foreign key to the user, if this declaration is associated with a registered user.
	/// Nullable to support potential future admin-seeded or imported declarations
	/// where only a bib number is known but no user account exists yet.
	/// Currently, all declarations are created via user self-declaration.
	/// </summary>
	public int? UserId { get; set; }

	/// <summary>
	/// Whether this is the user's first DLS attempt.
	/// </summary>
	public bool IsFirstDls { get; set; }

	/// <summary>
	/// Whether the user is going for kills (passing other runners).
	/// </summary>
	public bool IsGoingForKills { get; set; }

	/// <summary>
	/// Optional user comments (motivation, charity, etc.).
	/// </summary>
	public string? Comments { get; set; }

	/// <summary>
	/// Timestamp when this declaration was created.
	/// </summary>
	public DateTime CreatedAt { get; set; }

	/// <summary>
	/// Timestamp when this declaration was last modified.
	/// </summary>
	public DateTime ModifiedAt { get; set; }

	// Navigation properties

	/// <summary>
	/// The DLS race entry this declaration is for.
	/// </summary>
	public DlsRace DlsRace { get; set; } = null!;

	/// <summary>
	/// The user who made this declaration (if associated).
	/// </summary>
	public User? User { get; set; }
}
